### PR TITLE
Fix indefinite hang on OpenAPI schemas with cyclic model dependencies

### PR DESCRIPTION
**Pull request by kevin-paulson-mindbridge-ai**
We've been generating a model with the tool based on our OpenAPI schema, however since version 0.37.0 we've been unsuccessful in doing so. In the latest release (0.56.0) the generation was revealed to be stuck in `sort_data_models`. I waited up to an hour, I don't think it was ever going to complete. With this change it now works as expected.

I tried to follow https://datamodel-code-generator.koxudaxi.dev/development-contributing/ as best as I can, but let me know if you would like me to make any changes or if I should create an issue or anything else. Thanks!

## Summary by CodeRabbit

* **Bug Fixes**
  * Deterministic handling of circular dependencies in discriminated oneOf/allOf schemas.
  * Union type hints now omit duplicate members.
  * Safeguards added to skip invalid discriminator mappings when references are missing.

* **Tests**
  * Added JSON Schema and OpenAPI regression tests for discriminated oneOf/allOf circular dependency scenarios.

* **Chores**
  * Test tooling updated to include pytest-timeout.

### CODE REVIEW

**Review by ilovelinux — APPROVED** [review]
This PR implements cycle detection as it is usually done in graph algorithms (BFS, DFS, etc...).

LGTM! Thank you @kevin-paulson-mindbridge-ai 🙂

**Review by ilovelinux — CHANGES_REQUESTED** [review]
Hi @kevin-paulson-mindbridge-ai, I realized it would be great to add some generation test for schemas with cyclic references in order to track the generated output of this corner-case. 

See `tests/main/{openapi,jsonschema,graphql}/test_main_*.py` for examples. 🙂

**Review by ilovelinux — CHANGES_REQUESTED** [review]
I locally cherry-picked the last two commits (the ones that introduce the generation tests) to main. I expect them to fail (or hang) but they succeed instead.

Given remote:
```console
$ git remote add kevin-paulson-mindbridge-ai https://github.com/kevin-paulson-mindbridge-ai/datamodel-code-generator.git
$ git fetch kevin-paulson-mindbridge-ai
```

To replicate. exec from `main` branch:
```console
$ git cherry-pick 7786cf90867db72f716f1796ab071fefcffe4dc5
$ git cherry-pick 8d3ee3958ce8d9c595fdaa3d46a73bba2b54fa23
$ tox -e py314-parallel -- -k cyclic
```

This means that the cyclic model dependency added as test was already handled by the tool.

---

In fact, also the other tests (which implements the same cyclic model dependency corner-case) works on `main`.

```console
$ git switch --detach kevin-paulson-mindbridge-ai/slow_schema_generation_fix
$ git restore src/datamodel_code_generator/parser/base.py --source main
$ tox -e py314-parallel -- -k cyclic
```

---

@kevin-paulson-mindbridge-ai could you double-check the corner-case you have in your OpenAPI schema? As now, it looks like the one you proposed was already supported, therefore we cannot reproduce the issue, validate the fix & add regression tests 🙁

**Comment by kevin-paulson-mindbridge-ai** [comment]
> [...]
> 
> @kevin-paulson-mindbridge-ai could you double-check the corner-case you have in your OpenAPI schema? As now, it looks like the one you proposed was already supported, therefore we cannot reproduce the issue, validate the fix & add regression tests 🙁

My apologies, I neglected to validate that the new tests reproduced the issue before the change. I have updated the 3 tests such that they reproduce the issue before the change (they each ran at least 30 seconds before I killed them) and after the change they pass quickly (less than a second each). They look OK to me, but I am most familiar with OpenAPI and I am very unfamiliar with graphql.

File: tests/data/expected/main/openapi/cyclic_component_refs.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+class Model(BaseModel):
+    pass
+
+
+class Person(Team, Model):
+    name: str | None = None
+
+
+class Team(Person, Model):
+    name: str | None = None
```
[the remark this answers is not in the document]
Comment (kevin-paulson-mindbridge-ai) [reply tests/data/expected/main/openapi/cyclic_component_refs.py]: I'll see if I can create better test data.

**Comment by ilovelinux** [comment]
> My apologies, I neglected to validate that the new tests reproduced the issue before the change.

That's not a problem, I appreciate your effort. Reviews & double-checks are useful because of that :slightly_smiling_face: 

> I have updated the 3 tests such that they reproduce the issue before the change (they each ran at least 30 seconds before I killed them) and after the change they pass quickly (less than a second each). They look OK to me, but I am most familiar with OpenAPI and I am very unfamiliar with graphql.

I'll review it as soon as I'll get some free time :slightly_smiling_face: 

> (they each ran at least 30 seconds before I killed them)

Could you add a timeout as suggested by the first nitpick comment here: https://github.com/koxudaxi/datamodel-code-generator/pull/3078#pullrequestreview-4093826781?

File: tests/data/expected/main/openapi/cyclic_component_refs.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ilovelinux) [reply tests/data/expected/main/openapi/cyclic_component_refs.py]: @kevin-paulson-mindbridge-ai since you said those examples won't generate any output before[^1], and now they generate this output, I think this may been introduced by your bugfix.

That's an interesting corner-case. I think we should look into that before merging.

@kevin-paulson-mindbridge-ai IMHO if you can't fix that, consider adding an error to communicate to the user this is an (yet) unsupported corner-case, in order to avoid generating broken code. We may track that in another issue then and fix later. Finding the edges of the problem is the first step to solve it, after all.

CC: @koxudaxi WDYT? :slightly_smiling_face: 

[^1]: https://github.com/koxudaxi/datamodel-code-generator/pull/3078#issuecomment-4238086331

[the remark this answers is not in the document]
Comment (kevin-paulson-mindbridge-ai) [reply tests/data/expected/main/openapi/cyclic_component_refs.py]: I no longer think that this test data is fully reproducing the original issue and I don't think the fix is quite it either.

What caused me to look at this is we have a somewhat large schema available at test.mindbridge.ai/api/docs/mindbridge-openapi.json which produces with this change an about 20,000 line (valid) python file. Although with this change it will now generate the python file instead of never terminating:

1. The generated python file produces duplicate union members, this was not noticed as they were hidden by `ruff check --fix`
2. I'm having a lot of difficulty determining a test that reproduces the issue and still generates the valid Python expected.

I will plan to continue reviewing this PR later this week to see what I can discover.

**Comment by kevin-paulson-mindbridge-ai** [comment]
I think I've got things much closer now. I'll plan to come back to this to:

1) Add a test for the 1 missed line
2) Add similar tests for jsonschema and graphql

**Comment by kevin-paulson-mindbridge-ai** [comment]
@coderabbitai resume

**Comment by kevin-paulson-mindbridge-ai** [comment]
> I think I've got things much closer now. I'll plan to come back to this to:
> 
>     1. Add a test for the 1 missed line
> 
>     2. Add similar tests for jsonschema and graphql

This is now ready for review. Note that I don't believe graphql has the same issue due to the code paths followed and the nature of graphql.

**Review by koxudaxi — APPROVED** [review]

**Comment by koxudaxi** [comment]
@kevin-paulson-mindbridge-ai 
Thank you for the great PR. 
I reviewed it, and the previous concerns all appear to have been addressed, so I’ll merge it.
